File: cortex_web/apps/web/engine/policy.ts
// AD6 termination + verdict policy — port of scripts/cortex_policy.py.
//
// Per task k after each trial:
//   π_k    = Σ_i w_i · 1[ℓ_k^(i) > ℓ*_k]                 pass-mass
//   mcse_k = √( π_k(1−π_k) / ESS )                        MC error
//   R_k    = 1 − Var_post(ℓ_k) / Var_prior(ℓ_k)            info gate
// RESOLVED iff n_k ≥ N_min ∧ R_k ≥ R* and
//   PASS : π_k − Z·mcse_k ≥ 1 − α
//   FAIL : π_k + Z·mcse_k ≤ α
// Stop when all K resolved. finalize() maps leftover PENDING →
// REFER_BORDERLINE (gate opened) / REFER_UNINFORMATIVE (gate never opened).

import { ParticleState, TerminationPolicyName } from "./types";
import { ess } from "./particles";

export const VERDICT = {
  PASS: "PASS",
  FAIL: "FAIL",
  PENDING: "PENDING",
  REFER_BORDERLINE: "REFER_BORDERLINE",
  REFER_UNINFORMATIVE: "REFER_UNINFORMATIVE",
} as const;

// Paper-grade thresholds — synced with the v1.3.6 instrument freeze
// (data/INSTRUMENT_FREEZE_v1_3_6.json). Tighter α + higher N_min trade some
// questions for confidence; matches the desktop's shipped config.
export const DEFAULT_N_MIN = 20;
export const DEFAULT_R_STAR = 0.3;
export const DEFAULT_ALPHA = 0.05;
export const DEFAULT_Z = 2.0;

export interface PolicyResult {
  policyName: TerminationPolicyName;
  stop: boolean;
  stopReason: string;
  // Selection state is deliberately policy-neutral: AD6 emits its monotone
  // verdict locks; Precision emits fresh ACTIVE/ESTIMATE_COMPLETE/terminal
  // statuses. The selector only tests whether a domain remains active.
  selectionStates: string[];
  verdicts: string[];
  pi: number[];
  mcse: number[];
  R: number[];
  ess: number;
  domainStatuses?: string[];
  determinations?: string[];
  terminalReasons?: (string | null)[];
  streakCounts?: number[];
  diagnostics?: Record<string, unknown>;
}

export interface FinalPolicyResult {
  verdicts: string[];
  domainStatuses?: string[];
  determinations?: string[];
  terminalReasons?: (string | null)[];
  skillIntervals?: [number, number][];
  biasIntervals?: [number, number][];
}

export interface EngineTerminationPolicy {
  readonly name: TerminationPolicyName;
  readonly activeLabel: string;
  reset(K: number): void;
  evaluate(
    st: ParticleState,
    nPerTask: number[],
    telemetry?: Record<string, unknown>,
  ): PolicyResult;
  finalizeResult(ellStar: number[]): FinalPolicyResult;
  clone(): EngineTerminationPolicy;
  snapshot(): PolicySnapshot;
  restore(snapshot: PolicySnapshot): void;
}

export interface AD6PolicySnapshot {
  name: "ad6";
  verdicts: string[];
  lastR: number[];
}

export interface PrecisionPolicySnapshot {
  name: "precision_v1";
  statuses: string[];
  streaks: number[];
  terminalReasons: (string | null)[];
  bandAdministered: number[][];
  // Kept structurally typed at the policy boundary to avoid making the generic
  // policy interface depend on Precision's diagnostic implementation module.
  lastDiag: Record<string, unknown> | null;
}

export type PolicySnapshot = AD6PolicySnapshot | PrecisionPolicySnapshot;

export class AD6Policy implements EngineTerminationPolicy {
  readonly name = "ad6" as const;
  readonly activeLabel = VERDICT.PENDING;
  private ellStar: number[];
  private varPrior: number[];
  private nMin: number;
  private rStar: number;
  private alpha: number;
  private Z: number;
  private verdicts: string[];
  private lastR: number[];

  constructor(
    ellStar: number[],
    varPrior: number[],
    opts: { nMin?: number; rStar?: number; alpha?: number; Z?: number } = {},
  ) {
    this.ellStar = ellStar;
    this.varPrior = varPrior;
    this.nMin = opts.nMin ?? DEFAULT_N_MIN;
    this.rStar = opts.rStar ?? DEFAULT_R_STAR;
    this.alpha = opts.alpha ?? DEFAULT_ALPHA;
    this.Z = opts.Z ?? DEFAULT_Z;
    this.verdicts = new Array(ellStar.length).fill(VERDICT.PENDING);
    this.lastR = new Array(ellStar.length).fill(0);
  }

  // Var_prior from diag(Corr_l); ℓ* from cert_config v13.
  static fromInputs(
    ellStar: number[],
    corrL: number[][],
    opts = {},
  ): AD6Policy {
    const varPrior = corrL.map((row, i) => row[i]);
    return new AD6Policy(ellStar, varPrior, opts);
  }

  evaluate(st: ParticleState, nPerTask: number[]): PolicyResult {
    const { N, K, l, w } = st;
    const e = ess(w);
    const pi = new Array(K).fill(0);
    const mcse = new Array(K).fill(0);
    const R = new Array(K).fill(0);
    for (let k = 0; k < K; k++) {
      let piK = 0;
      let mu = 0;
      for (let n = 0; n < N; n++) {
        const lk = l[n * K + k];
        if (lk > this.ellStar[k]) piK += w[n];
        mu += w[n] * lk;
      }
      let varPost = 0;
      for (let n = 0; n < N; n++) {
        const d = l[n * K + k] - mu;
        varPost += w[n] * d * d;
      }
      pi[k] = piK;
      mcse[k] = Math.sqrt(Math.max(piK * (1 - piK), 0) / Math.max(e, 1));
      R[k] = 1 - varPost / this.varPrior[k];
    }
    // monotonic verdict update
    for (let k = 0; k < K; k++) {
      if (this.verdicts[k] !== VERDICT.PENDING) continue;
      if (nPerTask[k] < this.nMin || R[k] < this.rStar) continue;
      if (pi[k] - this.Z * mcse[k] >= 1 - this.alpha) this.verdicts[k] = VERDICT.PASS;
      else if (pi[k] + this.Z * mcse[k] <= this.alpha) this.verdicts[k] = VERDICT.FAIL;
    }
    this.lastR = R;
    const allResolved = this.verdicts.every((v) => v !== VERDICT.PENDING);
    return {
      policyName: this.name,
      stop: allResolved,
      stopReason: allResolved ? "all_resolved" : "continue",
      selectionStates: [...this.verdicts],
      verdicts: [...this.verdicts],
      pi,
      mcse,
      R,
      ess: e,
    };
  }

  reset(K: number): void {
    if (K !== this.ellStar.length) {
      throw new Error(`K=${K} != configured domains=${this.ellStar.length}`);
    }
    this.verdicts.fill(VERDICT.PENDING);
    this.lastR.fill(0);
  }

  finalize(): string[] {
    return this.verdicts.map((v, k) =>
      v !== VERDICT.PENDING
        ? v
        : this.lastR[k] >= this.rStar
          ? VERDICT.REFER_BORDERLINE
          : VERDICT.REFER_UNINFORMATIVE,
    );
  }

  finalizeResult(_ellStar: number[]): FinalPolicyResult {
    return { verdicts: this.finalize() };
  }

  // Deep copy of the monotonic verdict state — for speculative branch isolation
  // (engine/advance.ts). Params (ℓ*, varPrior, thresholds) are immutable and
  // shared; only verdicts + lastR are per-branch mutable.
  clone(): AD6Policy {
    const p = new AD6Policy(this.ellStar, this.varPrior, {
      nMin: this.nMin, rStar: this.rStar, alpha: this.alpha, Z: this.Z,
    });
    p.verdicts = this.verdicts.slice();
    p.lastR = this.lastR.slice();
    return p;
  }

  snapshot(): AD6PolicySnapshot {
    return {
      name: this.name,
      verdicts: this.verdicts.slice(),
      lastR: this.lastR.slice(),
    };
  }

  restore(snapshot: PolicySnapshot): void {
    if (snapshot.name !== this.name) throw new Error("AD6 policy snapshot mismatch");
    if (snapshot.verdicts.length !== this.ellStar.length
      || snapshot.lastR.length !== this.ellStar.length) {
      throw new Error("AD6 policy snapshot domain mismatch");
    }
    this.verdicts = snapshot.verdicts.slice();
    this.lastR = snapshot.lastR.slice();
  }
}

File: ad6-policy/src/ad6_policy/policy.py
from __future__ import annotations

import numpy as np

from cortex_termination_policy import StopDecision, TerminationPolicy

PASS = "PASS"
FAIL = "FAIL"
PENDING = "PENDING"
REFER_BORDERLINE = "REFER_BORDERLINE"
REFER_UNINFORMATIVE = "REFER_UNINFORMATIVE"

DEFAULT_N_MIN = 20
DEFAULT_R_STAR = 0.30
DEFAULT_ALPHA = 0.05
DEFAULT_Z = 2.0


class AD6Policy(TerminationPolicy):
    """AD6's monotone, cut-aware three-way verdict policy.

    For domain ``k`` the policy computes posterior mass above the supplied
    certification cut, a Monte-Carlo error guard, and posterior variance
    contraction. PASS and FAIL are locked once reached. Any unresolved domain
    is converted to a REFER reason when the session controller finalizes.

    Cut-vector loading intentionally lives outside this standalone package.
    """

    def __init__(
        self,
        ell_star,
        var_prior,
        *,
        n_min: int = DEFAULT_N_MIN,
        R_star: float = DEFAULT_R_STAR,
        alpha: float = DEFAULT_ALPHA,
        Z: float = DEFAULT_Z,
    ):
        self.ell_star = np.asarray(ell_star, dtype=float)
        self.var_prior = np.asarray(var_prior, dtype=float)
        if self.ell_star.shape != self.var_prior.shape:
            raise ValueError(
                f"ell_star ({self.ell_star.shape}) and var_prior "
                f"({self.var_prior.shape}) must align by task"
            )
        if (
            self.ell_star.ndim != 1
            or len(self.ell_star) == 0
            or np.any(~np.isfinite(self.ell_star))
            or np.any(~np.isfinite(self.var_prior))
            or np.any(self.var_prior <= 0.0)
        ):
            raise ValueError("ell_star and var_prior must be finite vectors")
        self.n_min = int(n_min)
        self.R_star = float(R_star)
        self.alpha = float(alpha)
        self.Z = float(Z)
        self._verdicts: list | None = None
        self._last_diag: dict | None = None

    @property
    def verdicts(self) -> list | None:
        return list(self._verdicts) if self._verdicts is not None else None

    @property
    def last_diagnostics(self) -> dict | None:
        return dict(self._last_diag) if self._last_diag is not None else None

    def reset(self, K: int) -> None:
        if len(self.ell_star) != K:
            raise ValueError(
                f"K={K} != len(ell_star)={len(self.ell_star)} — task mismatch"
            )
        self._verdicts = [PENDING] * K
        self._last_diag = None

    def __call__(self, state, telemetry, n_per_task, K) -> StopDecision:
        if self._verdicts is None:
            self.reset(K)
        w = np.asarray(state["w"], dtype=float)
        wsum = float(w.sum())
        w = w / wsum if wsum > 0 else np.full_like(w, 1.0 / len(w))
        skill = np.asarray(state["l"], dtype=float)
        ess = float(1.0 / float((w * w).sum()))

        pi = np.zeros(K)
        mcse = np.zeros(K)
        contraction = np.zeros(K)
        for k in range(K):
            ell_k = skill[:, k]
            pi_k = float((w * (ell_k > self.ell_star[k])).sum())
            mu_k = float((w * ell_k).sum())
            var_post = float((w * (ell_k - mu_k) ** 2).sum())
            pi[k] = pi_k
            mcse[k] = float(
                np.sqrt(max(pi_k * (1.0 - pi_k), 0.0) / max(ess, 1.0))
            )
            contraction[k] = 1.0 - var_post / float(self.var_prior[k])

        for k in range(K):
            if self._verdicts[k] != PENDING:
                continue
            if n_per_task[k] < self.n_min or contraction[k] < self.R_star:
                continue
            if pi[k] - self.Z * mcse[k] >= 1.0 - self.alpha:
                self._verdicts[k] = PASS
            elif pi[k] + self.Z * mcse[k] <= self.alpha:
                self._verdicts[k] = FAIL

        diagnostics = {
            "pi": pi.tolist(),
            "mcse": mcse.tolist(),
            "R": contraction.tolist(),
            "ess": ess,
            "verdicts": list(self._verdicts),
            "n_per_task": list(n_per_task),
        }
        self._last_diag = diagnostics
        stop = all(verdict != PENDING for verdict in self._verdicts)
        return StopDecision(
            stop=stop,
            stop_reason="all_resolved" if stop else "continue",
            verdicts=list(self._verdicts),
            diagnostics=diagnostics,
        )

    def finalize_verdicts(self) -> list:
        if self._verdicts is None:
            return []
        out = list(self._verdicts)
        contraction = (
            self._last_diag["R"] if self._last_diag else [0.0] * len(out)
        )
        for k, verdict in enumerate(out):
            if verdict == PENDING:
                out[k] = (
                    REFER_BORDERLINE
                    if contraction[k] >= self.R_star
                    else REFER_UNINFORMATIVE
                )
        return out
